### ISSUE TITLE
Potential data race on fHavePruned flag

### ISSUE THREAD

**Issue by BradSwain** [bitcoin/bitcoin#21627]
We detected a potential race on the global variable [`fHavePruned`][fHavePruned] with a static data race detection tool.

Because the report is from a static analysis tool,  we do not have a concrete execution or trace, but we have reviewed the report and are reasonably confident this variable can be accessed in parallel and is not guarded by any common locks.

We are not able to confirm 100% that this race is real, or what impacts it may have on the program, but we decided to report it just to be safe.  

#### Writing Thread
The flag can be set to true by the thread spawned at [`init.cpp:1883`](https://github.com/bitcoin/bitcoin/blob/9be7fe4849310884294669b019dd8300f69bc334/src/init.cpp#L1883)

https://github.com/bitcoin/bitcoin/blob/9be7fe4849310884294669b019dd8300f69bc334/src/init.cpp#L1883-L1885

This thread can call [`ThreadImport`][ThreadImport] -> [`LoadExternalBlockFile`][LoadExternalBlockFile] -> [`AcceptBlock`][AcceptBlock] -> [`FlushStateToDisk`][FlushStateToDisk]

And the [`FlushStateToDisk`][FlushStateToDisk] function can write to the [`fHavePruned`][fHavePruned] flag.

https://github.com/bitcoin/bitcoin/blob/9be7fe4849310884294669b019dd8300f69bc334/src/validation.cpp#L2311

[fHavePruned]: https://github.com/bitcoin/bitcoin/blob/9be7fe4849310884294669b019dd8300f69bc334/src/validation.cpp#L139
[ThreadImport]: https://github.com/bitcoin/bitcoin/blob/9be7fe4849310884294669b019dd8300f69bc334/src/init.cpp#L1884
[LoadExternalBlockFile]: https://github.com/bitcoin/bitcoin/blob/9be7fe4849310884294669b019dd8300f69bc334/src/init.cpp#L715
[AcceptBlock]: https://github.com/bitcoin/bitcoin/blob/9be7fe4849310884294669b019dd8300f69bc334/src/validation.cpp#L4717
[FlushStateToDisk]: https://github.com/bitcoin/bitcoin/blob/9be7fe4849310884294669b019dd8300f69bc334/src/validation.cpp#L3807

#### Reading Thread

The [`fHavePruned`][fHavePruned] flag can be read as part of the REST HTTP handler [`rest_block_extended`](https://github.com/bitcoin/bitcoin/blob/9be7fe4849310884294669b019dd8300f69bc334/src/rest.cpp#L300) which registered at [`rest.cpp:690`](
https://github.com/bitcoin/bitcoin/blob/9be7fe4849310884294669b019dd8300f69bc334/src/rest.cpp#L672-L693).

From there the handler can call [`rest_block`][rest_block] -> [`blockToJSON`][blockToJSON] -> [`IsBlockPruned`][isBlockPruned]

And [`IsBlockPruned`]() will read [`fHavePruned`][fHavePruned] 

https://github.com/bitcoin/bitcoin/blob/9be7fe4849310884294669b019dd8300f69bc334/src/validation.h#L1026

[isBlockPruned]: https://github.com/bitcoin/bitcoin/blob/9be7fe4849310884294669b019dd8300f69bc334/src/rpc/blockchain.cpp#L167
[blockToJSON]: https://github.com/bitcoin/bitcoin/blob/9be7fe4849310884294669b019dd8300f69bc334/src/rest.cpp#L285
[rest_block]: https://github.com/bitcoin/bitcoin/blob/9be7fe4849310884294669b019dd8300f69bc334/src/rest.cpp#L300

**Expected behavior**

No threads reading and writing to the same variable without synchronization.

**Actual behavior**

It appears possible for threads to read and write [`fHavePruned`][fHavePruned] in parallel.

**To reproduce**

May be difficult to reproduce consistently as we found the potential race through static analysis and data races can be non-deterministic.

**System information**

We analyzed commit 328aaac80 specifically, but the issue is still present at the time of posting this issue. 

For reference, here is a screenshot of the report generated by our tool ([Coderrect Scanner](https://coderrect.com/overview/)). I have summarized all the info from this report above.
![Coderrect Scanner Report](https://user-images.githubusercontent.com/25114061/113774750-4f7deb80-96ed-11eb-98e1-b7553a82f001.png)

**Comment by practicalswift** [comment]
@BradSwain Thanks a lot for reporting!

I love when our code base is analysed using new static analysis tools: the more the merrier! Did you find any other data race issues? Can you share your results in a gist or similar?

Unfortunately we're doing CI testing with quite a few ThreadSanitizer suppressions in https://github.com/bitcoin/bitcoin/blob/master/test/sanitizer_suppressions/tsan. Were you able to replicate any of those finds?

I think we should try to more clearly categorise our suppressions as "benign" vs. impactful: currently it is a mix of both which makes it hard to reason about the potential worst-case impact of each case. 

Last week a technical report on related work in the Firefox project was published: ["Eliminating Data Races in Firefox – A Technical Report"](https://hacks.mozilla.org/2021/04/eliminating-data-races-in-firefox-a-technical-report/). I think this makes sense:

> While benign data races do exist, we found (in agreement with previous work on this subject [1] [2]) that data races are very easily misclassified as benign. The reasons for this are clear: It is hard to reason about what compilers can and will optimize, and confirmation for certain “benign” data races requires you to look at the assembler code that the compiler finally produces.
> 
> Needless to say, this procedure is often much more time consuming than fixing the actual data race and also not future-proof. As a result, we decided that the ultimate goal should be a “no data races” policy that declares even benign data races as undesirable due to their risk of misclassification, the required time for investigation and the potential risk from future compilers (with better optimizations) or future platforms (e.g. ARM).

**Comment by BradSwain** [comment]
> Did you find any other data race issues?
> Were you able to replicate any of those finds?

We are still going through some of the reports. We will see if our tool can also find those races from the TSan suppression list.  
I will post anything else we find as a comment on this issue.

> Last week a technical report on related work in the Firefox project was published

I really like this report and I fully agree.

**Comment by amadeuszpawlik** [comment]
Don't [LoadExternalBlockFile](https://github.com/bitcoin/bitcoin/blob/9be7fe4849310884294669b019dd8300f69bc334/src/validation.cpp#L4700)  and [rest_block](https://github.com/bitcoin/bitcoin/blob/9be7fe4849310884294669b019dd8300f69bc334/src/rest.cpp#L251) both hold the `cs_main` though?

**Comment by BradSwain** [comment]
Thanks for taking a look @amadeuszpawlik 

My understanding may be wrong, but is the lock released at the end of this block (line 263)?

https://github.com/bitcoin/bitcoin/blob/9be7fe4849310884294669b019dd8300f69bc334/src/rest.cpp#L250-L263

**Comment by sipa** [comment]
@BradSwain Yes it is. `LOCK(cs_main);` locks for the duration of the scope it occurs in (like `std::unique_lock` etc).

**Comment by amadeuszpawlik** [comment]
@BradSwain now I see, I was staring at
[`rest_block -> IsBlockPruned`](https://github.com/bitcoin/bitcoin/blob/9be7fe4849310884294669b019dd8300f69bc334/src/rest.cpp#L258)
but it obviously doesn't hold for
[`rest_block -> blockToJSON -> IsBlockPruned`](https://github.com/bitcoin/bitcoin/blob/9be7fe4849310884294669b019dd8300f69bc334/src/rest.cpp#L285)
Thanks!

**Comment by MarcoFalke** [comment]
See also #17161

**Comment by pinheadmz** [comment]
I think this issue can be closed, `fHavePruned` seems to have been replaced with `chainman().m_blockman.m_have_pruned` (etc) and looks to me like all the setters (e.g. `FlushStateToDisk()`) have appropriate locks

**Comment by enirox001** [comment]
@pinheadmz are you still looking into this? Seems it's been a while since any progress has been made to it? Is it still 
relevant?

**Comment by pinheadmz** [comment]
I haven't been focusing on this, wanna take a stab at it?

**Comment by enirox001** [comment]
Sure, I can give it a go
